# IZV part3: Úkol 2: Test hypotézy
# Autor: Lukáš Kaprál (xkapra00)
Ověřovat s 95% jistotou následující dvě hypotézy
## Hypotéza 1:
**Na silnicích první třídy se při nehodách umíralo se stejnou pravděpodobností jako na
silnicích třetí třídy**

Import knihoven

In [23]:
import pandas as pd
import scipy as sp

Načteme dataframe s nehodami

In [24]:
df = pd.read_pickle('accidents.pkl.gz')
df

p1  p36    p37         p2a  weekday(p2a)   p2b  p6  p7  p8  \
0       002100160001    4     -1  2016-01-01             5    55   1   1   0   
1       002100160002    4     -1  2016-01-01             5   130   1   3   0   
2       002100160003    5     -1  2016-01-01             5   100   1   2   0   
3       002100160004    6     -1  2016-01-01             5   120   9   0   0   
4       002100160005    6     -1  2016-01-01             5  2560   2   0   0   
...              ...  ...    ...         ...           ...   ...  ..  ..  ..   
572929  190906210457    6     -1  2021-08-30             1   506   2   0   0   
572930  190906210458    6     -1  2021-08-30             1   752   2   0   0   
572931  190906210459    3   2099  2021-08-30             1  1505   1   3   0   
572932  190906210460    3   2099  2021-08-31             2  1245   0   0   0   
572933  190906210461    3  21029  2021-08-31             2  1834   3   0   4   

        p9  ...      l        n  o                       p       q       r  \
0        2  ...          711403     Souhlasnýsesměremúseku  Pomalý  554782   
1        2  ...                                                         -1   
2        2  ...                                                         -1   
3        2  ...                                                         -1   
4        2  ...                                                         -1   
...     ..  ...    ...      ... ..                     ...     ...     ...   
572929   2  ...         2611387                             Pomalý  560359   
572930   2  ...         2541503         Opačnýkesměruúseku  Pomalý  560383   
572931   2  ...                                                         -1   
572932   1  ...   2099   199764         Opačnýkesměruúseku  Pomalý  560286   
572933   1  ...  21029  3650728     Souhlasnýsesměremúseku  Pomalý  560324   

             s                   t  p5a  region  
0       451622  GN_V0.1UIR-ADR_410    1     PHA  
1           -1                        1     PHA  
2           -1                        1     PHA  
3           -1                        1     PHA  
4           -1                        1     PHA  
...        ...                 ...  ...     ...  
572929      -1  GN_V0.1UIR-ADR_410    1     KVK  
572930  550426  GN_V0.1UIR-ADR_410    1     KVK  
572931      -1                        1     KVK  
572932  553654  GN_V0.1UIR-ADR_410    1     KVK  
572933      -1  GN_V0.1UIR-ADR_410    2     KVK  

[572934 rows x 65 columns]

Vyfiltrujeme pouze nehody na silnicích 1. a 3. třídy

In [25]:
df = df[df["p36"].isin([1, 3])]
df

p1  p36    p37         p2a  weekday(p2a)   p2b  p6  p7  p8  \
7594    002100167619    1     -1  2016-05-05             4  1135   1   2   0   
10530   003100160575    1    102  2016-06-19             0  1905   1   2   0   
10532   003100160577    1      0  2016-06-19             0  1915   3   0   4   
13362   003100163418    1      0  2016-08-08             1  1355   1   2   0   
21478   004100161576    1     -1  2016-12-09             5  1640   2   0   0   
...              ...  ...    ...         ...           ...   ...  ..  ..  ..   
572914  190906210442    3  21036  2021-08-23             1  1238   3   0   9   
572922  190906210450    3  21810  2021-08-26             4  1704   1   1   0   
572931  190906210459    3   2099  2021-08-30             1  1505   1   3   0   
572932  190906210460    3   2099  2021-08-31             2  1245   0   0   0   
572933  190906210461    3  21029  2021-08-31             2  1834   3   0   4   

        p9  ...      l        n  o                       p       q       r  \
7594     2  ...     R1  1397508     Souhlasnýsesměremúseku  Pomalý  554782   
10530    2  ...    102   312370     Souhlasnýsesměremúseku  Pomalý  554782   
10532    2  ...     R1   293620     Souhlasnýsesměremúseku  Pomalý  554782   
13362    2  ...         2374326     Souhlasnýsesměremúseku  Pomalý  554782   
21478    2  ...                                                         -1   
...     ..  ...    ...      ... ..                     ...     ...     ...   
572914   1  ...  21036  2245808         Opačnýkesměruúseku  Pomalý  560341   
572922   2  ...  21810  3343415     Souhlasnýsesměremúseku  Pomalý  560677   
572931   2  ...                                                         -1   
572932   1  ...   2099   199764         Opačnýkesměruúseku  Pomalý  560286   
572933   1  ...  21029  3650728     Souhlasnýsesměremúseku  Pomalý  560324   

             s                   t  p5a  region  
7594    709450  GN_V0.1UIR-ADR_410    1     PHA  
10530   481548  GN_V0.1UIR-ADR_410    1     PHA  
10532   709450  GN_V0.1UIR-ADR_410    1     PHA  
13362   506001  GN_V0.1UIR-ADR_410    1     PHA  
21478       -1  GN_V0.1UIR-ADR_410    1     PHA  
...        ...                 ...  ...     ...  
572914      -1  GN_V0.1UIR-ADR_410    2     KVK  
572922      -1  GN_V0.1UIR-ADR_410    1     KVK  
572931      -1                        1     KVK  
572932  553654  GN_V0.1UIR-ADR_410    1     KVK  
572933      -1  GN_V0.1UIR-ADR_410    2     KVK  

[153329 rows x 65 columns]

Vytvoříme nový sloupec, kam uložíme True/False podle toho, zda při nehodě někdo umřel

In [26]:
df["died"] = df["p13a"] > 0
df

/tmp/ipykernel_383/2265180932.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["died"] = df["p13a"] > 0


p1  p36    p37         p2a  weekday(p2a)   p2b  p6  p7  p8  \
7594    002100167619    1     -1  2016-05-05             4  1135   1   2   0   
10530   003100160575    1    102  2016-06-19             0  1905   1   2   0   
10532   003100160577    1      0  2016-06-19             0  1915   3   0   4   
13362   003100163418    1      0  2016-08-08             1  1355   1   2   0   
21478   004100161576    1     -1  2016-12-09             5  1640   2   0   0   
...              ...  ...    ...         ...           ...   ...  ..  ..  ..   
572914  190906210442    3  21036  2021-08-23             1  1238   3   0   9   
572922  190906210450    3  21810  2021-08-26             4  1704   1   1   0   
572931  190906210459    3   2099  2021-08-30             1  1505   1   3   0   
572932  190906210460    3   2099  2021-08-31             2  1245   0   0   0   
572933  190906210461    3  21029  2021-08-31             2  1834   3   0   4   

        p9  ...        n  o                       p       q       r       s  \
7594     2  ...  1397508     Souhlasnýsesměremúseku  Pomalý  554782  709450   
10530    2  ...   312370     Souhlasnýsesměremúseku  Pomalý  554782  481548   
10532    2  ...   293620     Souhlasnýsesměremúseku  Pomalý  554782  709450   
13362    2  ...  2374326     Souhlasnýsesměremúseku  Pomalý  554782  506001   
21478    2  ...                                                  -1      -1   
...     ..  ...      ... ..                     ...     ...     ...     ...   
572914   1  ...  2245808         Opačnýkesměruúseku  Pomalý  560341      -1   
572922   2  ...  3343415     Souhlasnýsesměremúseku  Pomalý  560677      -1   
572931   2  ...                                                  -1      -1   
572932   1  ...   199764         Opačnýkesměruúseku  Pomalý  560286  553654   
572933   1  ...  3650728     Souhlasnýsesměremúseku  Pomalý  560324      -1   

                         t  p5a  region   died  
7594    GN_V0.1UIR-ADR_410    1     PHA  False  
10530   GN_V0.1UIR-ADR_410    1     PHA  False  
10532   GN_V0.1UIR-ADR_410    1     PHA  False  
13362   GN_V0.1UIR-ADR_410    1     PHA  False  
21478   GN_V0.1UIR-ADR_410    1     PHA  False  
...                    ...  ...     ...    ...  
572914  GN_V0.1UIR-ADR_410    2     KVK  False  
572922  GN_V0.1UIR-ADR_410    1     KVK  False  
572931                        1     KVK  False  
572932  GN_V0.1UIR-ADR_410    1     KVK  False  
572933  GN_V0.1UIR-ADR_410    2     KVK  False  

[153329 rows x 66 columns]

Tvorba crosstab

In [27]:
ct = pd.crosstab(df["p36"], df["died"])
ct

died  False  True 
p36               
1     78618    911
3     73352    448

Spočítáme 𝜒2

In [28]:
sp.stats.chi2_contingency(ct)[1]

3.5395243450138555e-29

Výsledná hodnota 3.54*10<sup>-29</sup> je menší, než 0.05. Hypotéza neplatí 

## Hypotéza 1:
**Při nehodách vozidel značky Škoda je škoda na vozidle nižší než při nehodách vozidel Audi**

Vyfiltrujeme pouze vozidla značky Škoda a Audi (p45a 2=AUDI, 39=ŠKODA)

In [30]:
df = df[df["p45a"].isin([2, 39])]
df

p1  p36    p37         p2a  weekday(p2a)   p2b  p6  p7  p8  \
10530   003100160575    1    102  2016-06-19             0  1905   1   2   0   
50445   002100184560    1     -1  2018-03-20             2  1100   1   2   0   
117991  002100162546    1      0  2016-02-16             2  1110   1   2   0   
117994  002100162631    1      0  2016-02-17             3  1010   1   4   0   
119398  010106160002    3   1256  2016-01-04             1  2005   3   0   6   
...              ...  ...    ...         ...           ...   ...  ..  ..  ..   
572838  190906210366    3   2099  2021-07-15             4  1932   3   0   9   
572845  190906210373    3   2099  2021-07-19             1   913   1   2   0   
572868  190906210396    3  21036  2021-08-02             1  1605   3   0   4   
572888  190906210416    3   2099  2021-08-02             1  1520   2   0   0   
572931  190906210459    3   2099  2021-08-30             1  1505   1   3   0   

        p9  ...        n  o                       p       q       r       s  \
10530    2  ...   312370     Souhlasnýsesměremúseku  Pomalý  554782  481548   
50445    2  ...  2775387     Souhlasnýsesměremúseku  Pomalý  554782  467359   
117991   2  ...  1871811     Souhlasnýsesměremúseku  Pomalý  554782  709450   
117994   2  ...                                                  -1      -1   
119398   2  ...  1013486     Souhlasnýsesměremúseku  Pomalý  529931      -1   
...     ..  ...      ... ..                     ...     ...     ...     ...   
572838   2  ...   798308     Souhlasnýsesměremúseku  Pomalý  560537  552526   
572845   2  ...  3089393     Souhlasnýsesměremúseku  Pomalý  560642      -1   
572868   2  ...  1425538     Souhlasnýsesměremúseku  Pomalý  560588      -1   
572888   2  ...   199919     Souhlasnýsesměremúseku  Pomalý  560286  553654   
572931   2  ...                                                  -1      -1   

                         t  p5a  region   died  
10530   GN_V0.1UIR-ADR_410    1     PHA  False  
50445   GN_V0.1UIR-ADR_410    1     PHA  False  
117991  GN_V0.1UIR-ADR_410    2     STC  False  
117994                        2     STC  False  
119398  GN_V0.1UIR-ADR_410    2     STC  False  
...                    ...  ...     ...    ...  
572838  GN_V0.1UIR-ADR_410    1     KVK  False  
572845  GN_V0.1UIR-ADR_410    2     KVK  False  
572868  GN_V0.1UIR-ADR_410    2     KVK  False  
572888  GN_V0.1UIR-ADR_410    1     KVK  False  
572931                        1     KVK  False  

[41175 rows x 66 columns]

Tvorba crosstab, p53 = škoda na vozidle ve stokorunách

In [31]:
ct = pd.crosstab(df["p53"], df["p45a"])
ct

p45a   2     39
p53            
0      98  1146
1       3    45
2       0    14
3       0     5
4       0     1
...    ..   ...
9000    1     1
10000   3     5
12000   2     0
15000   2     0
20000   1     0

[138 rows x 2 columns]

Pomocí vhodného testu spočítáme, zda hypotéza platí

In [37]:
sp.stats.ttest_ind(ct[2], ct[39], equal_var=False)


Ttest_indResult(statistic=-3.3668242256670027, pvalue=0.0009836103135760946)

Výsledná hodnota 0.0009836 je menší, než 0.05. Hypotéza tedy neplatí.